sklearn/linear_model/_base.py

```diff
@@ -685,7 +685,7 @@ def rmatvec(b):
                 self._residues = np.vstack([out[3] for out in outs])
         else:
             self.coef_, self._residues, self.rank_, self.singular_ = \
-                linalg.lstsq(X, y)
+                linalg.lstsq(X, y, check_finite=False)
```

Comment:
I think we should be careful here. I don't see any safeguard in `_rescale_data`. Indeed, this operation take the `sqrt` of the sample weight. If we have negative `sample_weight`, we will get some `NaN` and without the `check_finite` we will have a segfault.

@ogrisel @agramfort Should we make sure that we should not have negative `sample_weight` here (in which case `check_finite=False` would be safe I think)?

Reply:
In `_rescale_data`, can result in non-finite values with finite input:

```python
from sklearn.linear_model._base import _rescale_data
import numpy as np

n_samples = 4
n_features = 4
LARGE = 1e-100 * np.finfo(np.float64).max   # ~1e208

sample_weight = np.full(n_samples, fill_value=LARGE)
X = np.full((n_samples, n_features), fill_value=LARGE)
y = np.full(n_samples, fill_value=LARGE)

_rescale_data(X, y, sample_weight)
# (array([[inf, inf, inf, inf],
#        [inf, inf, inf, inf],
#        [inf, inf, inf, inf],
#        [inf, inf, inf, inf]]),
#  array([inf, inf, inf, inf]))
```

Reply:
yes sample_weight should be positive numbers

Reply:
@ogrisel @agramfort Should we make sure that we should not have negative sample_weight here (in which case check_finite=False would be safe I think)?

Even when `check_finite=True`, making sure that sample wights are positive is helpful to generate more meaningful error messages.

This should probably be done as part of `sklearn.utils.validation._check_sample_weight`, probably in a dedicated pull request. We could have a `check_positive=True` options to disable this check in specific places of scikit-learn if needed but I think it should be enabled by default.

Reply:
https://github.com/scikit-learn/scikit-learn/pull/20152#discussion_r641550347 is not about negative weights though. So maybe we should not do `linalg.lstsq(X, y, check_finite=False)` at all, irrespective of the positive `sample_weight` check.